### PR TITLE
Fix SubplotGrid indexing and enhance legend placement with 'ref' argument

### LINKED ISSUE #459: Thoughts on PR #418 about legend placement
I like what you have done in PR #418. But there's a little more I'd like to discuss if you don't mind. Please excuse me if I'm mistaken.

I've been exploring the `ax/row(s)/col(s)/span` arguments of legend. These arguments are useful when someone wants the legend placed based on specific axes rather than the whole figure. However, when using this feature, two questions naturally arise: which axes' elements to include and where to position the legend. This leads to some confusion. I don't have definite answers, but I'd like to highlight this for your consideration.

Take the following code (without the fig.legend part) as an example, and let's assume `loc="bottom` throughout the discussion.

```python
import ultraplot as pplt
fig, axs = pplt.subplots(ncols=4, nrows=4, share='all', width='10cm')
axs.format(yticklabels=[], xticklabels=[])
cycle = pplt.Cycle("default", N=16)
for n, ax in enumerate(axs):
    ax.text(.5, .5,f'ax:{n}\ncols:{n%4+1}', transform=ax.transAxes, ha='center',va='center')
    ax.scatter(.5, .25, transform=ax.transAxes, color=cycle.get_next()['color'], label=f"ax:{n}")

# Now the fig.legend() part.
```

1. `fig.legend(ax=axs[4:6], loc='b', span=(1,2))` works as expected:

    1. `axs[4:6]` → includes elements from `axs[4]` and `axs[5]`.
    2. `span=(1,2)` → merges them into a single legend and places it below columns 1 and 2, i.e., below `axs[4]` and `axs[5]`.

2. `fig.legend(ax=axs[4:6], loc="bottom", span=(3,4))` works, though seems odd but makes sense with complex layout. It includes elements from `axs[4]` and `axs[5]` but places the legend below columns 3 and 4, i.e., below `axs[6]` and `axs[7]`. What about other rows? Since `ax` and `cols` can be disrelevant, then how to put legend under 1st row, i.e., below `axs[2]` and `axs[3]`

Currently, `ax` accepts a zero‑based slice but not an array. [4:7] is valid, while [4,5] is not. Meanwhile, `span/cols` accept a one‑based tuple but not a slice or array. In the example above, for the second row, cols=(1,3) essentially corresponds to axs[4:7].

If `ax` controls which elements are included, it should accept a scalar, slice, or list. And if span=(1,3) means “from column 1 to column 3,” that makes sense. But why does axs start at 0 while cols start at 1?

Trying `fig.legend(ax=axs[0,3,6], loc='bottom', span=(3,4)) # or axs[[0,3,6]]` results in only `axs[3]` being included, which doesn't seem right.

### PR DESCRIPTION
**Pull request by cvanelteren**
Closes #459, #460

This PR addresses issues with `SubplotGrid` indexing and significantly enhances the flexibility and intuitiveness of legend placement.

**1. SubplotGrid 1D List/Array Indexing**
*   **Fix:** `SubplotGrid` (returned by `subplots`) now supports 1D list or numpy array indexing (e.g., `axs[[0, 5]]`). This returns a new `SubplotGrid` containing only the selected axes, whereas previously it was misinterpreted as 2D indexing or raised errors.

**2. Decoupled Legend Placement (`ref` argument)**
*   **Feature:** Added a `ref` keyword argument to `fig.legend` (and `fig.colorbar`).
    *   `ax`: Specifies the **content** source (handles/labels).
    *   `ref`: Specifies the **location** anchor.
    *   This allows patterns like `fig.legend(ax=axs[1, :], ref=axs[0, :], loc='bottom')` to place a legend using data from the second row *below the first row*.
*   **Intelligent Placement & Inference:**
    *   **Implicit Span:** If `ref` (or `ax` if `ref` is missing) is a list of axes (e.g., `axs[[0, 1]]`) and no explicit span (`rows`/`cols`) is provided, UltraPlot now **automatically infers the span** based on the legend location. For example, `fig.legend(ax=axs[[0, 1]], loc='bottom')` will create a single legend spanning the width of both axes.
    *   **Edge Anchoring:** The legend is automatically anchored to the appropriate outer edge of the `ref` group (e.g., for `loc='right'`, it anchors to the rightmost axis in the group; for `loc='bottom'`, the bottommost).

**Documentation & Tests**
*   Added a new tutorial section "Decoupling legend content and location" to `docs/colorbars_legends.py`.
*   Added comprehensive tests in `ultraplot/tests/test_gridspec.py` (slicing) and `ultraplot/tests/test_legend.py` (ref argument, span inference).

**API Changes**
*   No breaking API changes. The `ref` argument is optional, and standard `fig.legend` usage remains compatible.

image

### CODE REVIEW

File: docs/colorbars_legends.py [hunk 1]
```diff
@@ -469,3 +469,44 @@
...
+#
+# Decoupling legend content and location
+# --------------------------------------
+#
+# Sometimes you may want to generate a legend using handles from specific axes
+# but place it relative to other axes. In UltraPlot, you can achieve this by passing
+# both the `ax` and `ref` keywords to :func:`~ultraplot.figure.Figure.legend`
+# (or :func:`~ultraplot.figure.Figure.colorbar`). The `ax` keyword specifies the
+# axes used to generate the legend handles, while the `ref` keyword specifies the
+# reference axes used to determine the legend location.
```
[the remark this answers is not in the document]
Comment (cvanelteren) [reply docs/colorbars_legends.py:483]: Handled now! Thanks

File: ultraplot/tests/test_legend.py [hunk 2]
```diff
@@ -529,3 +529,38 @@ def test_legend_explicit_handles_labels_override_auto_collection():
     assert leg is not None
     assert len(leg.get_texts()) == 1
     assert leg.get_texts()[0].get_text() == "custom_label"
+
+
+def test_legend_ref_argument():
+    """Test using 'ref' to decouple legend location from content axes."""
+    fig, axs = uplt.subplots(nrows=2, ncols=2)
+    axs[0, 0].plot([], [], label="line1")  # Row 0
+    axs[1, 0].plot([], [], label="line2")  # Row 1
```
[the remark this answers is not in the document]
Comment (cvanelteren) [reply ultraplot/tests/test_legend.py:538]: ignored.

File: ultraplot/figure.py [hunk 3]
```diff
@@ -2700,27 +2700,125 @@ def legend(
...
+                not has_span
+                and np.iterable(loc_ax)
+                and not isinstance(loc_ax, (str, maxes.Axes))
+            ):
+                loc_trans = _translate_loc(loc, "legend", default=rc["legend.loc"])
+                side = (
+                    loc_trans
+                    if loc_trans in ("left", "right", "top", "bottom")
+                    else None
+                )
```
[the remark this answers is not in the document]
Comment (cvanelteren) [reply ultraplot/figure.py:2829]: Good -- implemented

**Comment by gepcel** [comment]
I've had some simple tests, very brilliant work. Now the logic is very simple and very clear.

It already satisfied for pretty much everything. Bringing up those weird edge cases is kind of like looking for faults, unless you're really into making it perfect and still care about that stuff.

**Comment by cvanelteren** [comment]
Failing image is shape difference fyi

**Comment by cvanelteren** [comment]
> I've had some simple tests, very brilliant work. Now the logic is very simple and very clear.
> 
> 
> 
> It already satisfied for pretty much everything. Bringing up those weird edge cases is kind of like looking for faults, unless you're really into making it perfect and still care about that stuff.

Any cases we should be aware of?

**Comment by gepcel** [comment]
> Any cases we should be aware of?

Example 1
```python
# %% test ultraplot legend
import ultraplot as pplt
fig, axs = pplt.subplots(ncols=4, nrows=4, share='all', width='10cm')
axs.format(yticklabels=[], xticklabels=[])
cycle = pplt.Cycle("default", N=16)
for n, ax in enumerate(axs):
    ax.text(.5, .5, f'ax:{n}\ncols:{n%4+1}', transform=ax.transAxes, ha='center',va='center')
    ax.scatter(.5, .25, transform=ax.transAxes, color=cycle.get_next()['color'], label=f"ax{n}")
fig.legend(ax=axs[1, 1:], loc="b", ref=axs[2,:-1])
fig.legend(ax=axs[[3,5,8,12]], loc="b", ref=axs[0,1:-1])
fig.legend(ax=axs[-2:], loc='l', ref=axs[1:3, 3], ncols=1)
```

**Example 2**

```python
# %% test ultraplot legend
import ultraplot as pplt
fig, axs = pplt.subplots(ncols=4, nrows=4, share='all', width='10cm')
axs.format(yticklabels=[], xticklabels=[])
cycle = pplt.Cycle("default", N=16)
for n, ax in enumerate(axs):
    ax.text(.5, .5, f'ax:{n}\ncols:{n%4+1}', transform=ax.transAxes, ha='center',va='center')
    ax.scatter(.5, .25, transform=ax.transAxes, color=cycle.get_next()['color'], label=f"ax{n}")
fig.legend(ax=axs[-2:], loc='l', ref=axs[1:3, 3], ncols=1)
fig.legend(ax=axs[1, 1:], loc="b", ref=axs[2,:-1])
fig.legend(ax=axs[[3,5,8,12]], loc="b", ref=axs[0,1:-1])
```

The only difference is the last line moves two lines up. It not something I need, just play around for fun.

**Comment by cvanelteren** [comment]
I see. The legend does take up space to prevent overlap. For outside legends it is more a feature than a bug.

**Comment by gepcel** [comment]
And there's also overlaps between legend and axs when space of axs is small. Take `fig, axs = pplt.subplots(ncols=2, nrows=2, space='1mm')` for an example.

**Comment by cvanelteren** [comment]
It could be that these will be fixed by #430 -- this will replace the layout mechanism with one that is constrained based.

**Review by beckermr — APPROVED** [review]

**Comment by cvanelteren** [comment]
@gepcel I assume you are also approving the changes? If so can you go ahead and provide a review and approve the changes.
